qiskit/utils/deprecation.py

```diff
@@ -275,24 +293,36 @@ def wrapper(*args, **kwargs):
 
 
 def _maybe_warn_and_rename_kwarg(
-    func_name: str,
+    args: tuple[Any, ...],
     kwargs: Dict[str, Any],
     *,
-    old_arg: str,
+    func_name: str,
+    original_func_co_varnames: tuple[str, ...],
+    old_arg_name: str,
     new_alias: Optional[str],
     warning_msg: str,
     category: Type[Warning],
     predicate: Optional[Callable[[Any], bool]],
 ) -> None:
-    if old_arg not in kwargs:
+    # In Python 3.10+, we should set `zip(strict=False)` (the default). That is, we want to
+    # stop iterating once `args` is done.
+    arg_names_to_values = {name: val for val, name in zip(args, original_func_co_varnames)}
+    arg_names_to_values.update(kwargs)
```

Comment:
Come to think of it, it'd be good to put in a couple of tests that deprecations with variadic arguments (`*args` and/xor `**kwargs`) play nicely with all the machinery in these decorators.

Reply:
> I'm also getting a little concerned about the run-time effects on performance of these decorators now - there's really very non-trivial overhead in calculating all this stuff about rebinding, even for very simple "don't pass this argument" deprecations that were previously just a pointer comparison in the happy path (if x is not None).

There's a big downside here to code complexity & it being easy for users not to realize they need to set it. But, we could have an argument `@deprecate_arg(check_positional=True)`, with default to `True`?

This is tricky because I'm not seeing an optimal design that can handle everything we want. We either get full robustness with checking positional args or have less performance impact. A reminder that the status quo of `warnings.warn` had issues too with no warning on explicitly setting `bad_arg=None` (https://github.com/Qiskit/qiskit-terra/issues/9992#issuecomment-1515157692)

Reply:
Thanks for the thoughtful review! Test added and your suggestion indeed fixed things.